test-data/unit/check-based-callable.test

```diff
@@ -0,0 +1,275 @@
+[case testCallableSyntax]
+c: "(str) -> int"
+reveal_type(c)  # N: Revealed type is "(str) -> int"
+f: "def (str) -> int"
+reveal_type(f)  # N: Revealed type is "def (str) -> int"
+[builtins fixtures/tuple.pyi]
+
+
+[case testCallableSyntaxBreak]
+# ensure things that look a lot like Callable syntax are handled as not that
+from typing import Literal
+t: "(Literal['->'])"
+reveal_type(t)  # N: Revealed type is "'->'"
+[builtins fixtures/tuple.pyi]
+[typing fixtures/typing-medium.pyi]
+
+
+[case testCallableIsntFunction]
+c: "(str) -> int"
+c.__name__  # E: "(str) -> int" has no attribute "__name__"  [attr-defined]
+f: "def (str) -> int"
+reveal_type(f.__name__)  # N: Revealed type is "str"
+[builtins fixtures/tuple.pyi]
+[typing fixtures/typing-medium.pyi]
+
+
+[case testFunctionTypeSubtypesCallable]
+c: "(str) -> int"
+f: "def (str) -> int" = c  # E: Incompatible types in assignment (expression has type "(str) -> int", variable has type "def (str) -> int")  [assignment]
+c2: "(str) -> int" = f
+[builtins fixtures/tuple.pyi]
+
+
+[case testFunctiontypeInferred]
+def f(): ...
+reveal_type(f)  # N: Revealed type is "def () -> None"
+[builtins fixtures/tuple.pyi]
+
+
+[case testRuntimeFunctionType]
+from types import FunctionType
+import future
+
+a: FunctionType[[int], str]  # E: Type parameters for "FunctionType" requires __future__.annotations or quoted types  [valid-type]
+FunctionType[[int], str]  # E: The type "type[FunctionType]" is not generic and not indexable  [misc]
+[file future.py]
+from __future__ import annotations
+from types import FunctionType
+a: FunctionType[[int], str]
+FunctionType[[int], str]  # E: The type "type[FunctionType]" is not generic and not indexable  [misc]
+[builtins fixtures/tuple.pyi]
+
+
+[case testFunctionTypeOnInstance]
+from typing import ClassVar
+class A:
+    a: ClassVar["def (A) -> int"]
+a = A()
+A.a(a)
+a.a()
+[builtins fixtures/tuple.pyi]
+
+
+[case testMethodType]
+import types
+class A:
+    def f(self): ...
+reveal_type(A().f)  # N: Revealed type is "_NamedCallable & () -> None"
```

Comment:
since `_NamedCallable` extends `Callable`, wouldn't this intersection turn it into an overload of `Callable[..., Any] & Callable[[], None]`, which would incorrectly allow you to call the function with any arguments

Reply:
No it wouldn't, I checked :)